src/twisted/names/client.py

```diff
@@ -217,13 +217,26 @@ def _connectedProtocol(self, interface=''):
         Return a new L{DNSDatagramProtocol} bound to a randomly selected port
         number.
         """
+        failures = 0
         proto = dns.DNSDatagramProtocol(self, reactor=self._reactor)
+
         while True:
             try:
                 self._reactor.listenUDP(dns.randomSource(), proto,
                                         interface=interface)
-            except error.CannotListenError:
-                pass
+            except error.CannotListenError as e:
+                failures += 1
+
+                if (hasattr(e.socketError, "errno") and
+                   e.socketError.errno == errno.EMFILE):
+                    # We've run out of file descriptors. Stop trying.
+                    raise
+
+                if failures >= 1000:
```

Comment:
This seems like an odd idiom to repeat a thing.  Why not `for port in (randomPort() for each in six.range(1000)):` with a `return proto` in the `try` block?